src/cargo/util/semver_ext.rs

```diff
@@ -49,22 +51,48 @@ impl OptVersionReq {
                     cmp.op == Op::Exact && cmp.minor.is_some() && cmp.patch.is_some()
                 }
             }
+            OptVersionReq::Locked(..) => true,
+        }
+    }
+
+    pub fn lock_to(&mut self, version: &Version) {
+        assert!(self.matches(version), "cannot lock {} to {}", self, version);
+        use OptVersionReq::*;
+        let version = version.clone();
+        *self = match self {
+            Any => Locked(version, VersionReq::STAR),
+            Req(req) => Locked(version, req.clone()),
+            Locked(_, req) => Locked(version, req.clone()),
+        };
+    }
+
+    pub fn is_locked(&self) -> bool {
+        matches!(self, OptVersionReq::Locked(..))
+    }
+
+    /// Gets the version to which this req is locked, if any.
+    pub fn locked_version(&self) -> Option<&Version> {
+        match self {
+            OptVersionReq::Locked(version, _) => Some(version),
+            _ => None,
         }
     }
 
     pub fn matches(&self, version: &Version) -> bool {
         match self {
             OptVersionReq::Any => true,
             OptVersionReq::Req(req) => req.matches(version),
+            OptVersionReq::Locked(v, _) => VersionReq::exact(v).matches(version),
```

Comment:
Can thes just be `v == version`? Or are there problems with the metadata part? Are there test to make sure we don't accidentally switch to the "simpler code" if it does not work?

Reply:
Yes the metadata is counted in when comparing two `Version`s. We can reimplement the logic from our side since it is simple, but I am uncertain whether it is appropriate to do that.

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=1c01a049bc0aba5850ff8e048c5536cf

Reply:
I've updated the logic with 4ecb543. The test cases are shamelessly copied from [semver crate](https://github.com/dtolnay/semver/blob/2117ba6d8e7a524b671cee65883a19c834eb901f/tests/test_version_req.rs#L38-L62). Do you think it works?